File: LC/[1105] Filling Bookcase Shelves.cpp
class Solution {
public:
    int minHeightShelves(vector<vector<int>>& books, int shelfWidth) {
        //when do you need to insert a dummy head?
        //when ii has something to do with ii-1
        books.insert(books.begin(), {0,0});
        int N = books.size();
        vector<int> dp(N, INT_MAX);
        dp[0]=0; //always need to initialize DP

        /* DP Type II - Basic */
        for(int ii=1; ii<N; ++ii){          //DP type II - Basic 
            int max_height=0;
            int total_width=0;
            for(int jj=ii; jj >=1; --jj){ //try all possibility that can be on this shelf ending with books[ii]
                max_height=max(max_height, books[jj][1]);
                total_width += books[jj][0];
                if(total_width > shelfWidth) break;
                dp[ii]=min(dp[ii], dp[jj-1]+max_height);
            }
        }
        return dp[N-1];
    }
};
/*
dp[ii] = min height of bookshelf @ iith book
dp[ii]=dp[jj]+max(height @ current shelf)
*/

File: LC/[1105] Filling Bookcase Shelves.java
class Solution {
    public int minHeightShelves(int[][] books, int shelfWidth) {
        int N = books.length;
        int [] dp = new int[N+1];
        int k = shelfWidth;

        //insert dummy
        int[][] books2 = new int[N+1][2];//each ii is (thickness, height)
        for(int ii=0; ii<N; ++ii){
            books2[ii+1]=books[ii];
        }

        books = books2;
        //ii=0 (dummy head)
        dp[0]=0;
        //initialize dp
        for(int ii=1; ii<=N; ++ii) dp[ii] = Integer.MAX_VALUE/2;

        //populate dp
        for(int ii=1; ii<=N; ++ii){//for each book
            int curr_thickness=0;
            int mx=0;
            for(int ss=ii; ss>=1; --ss){
                curr_thickness += books[ss][0];
                mx = Math.max(mx, books[ss][1]);
                if(curr_thickness > k) break;
                dp[ii] = Math.min(dp[ii], dp[ss-1]+mx);
            }

        }

        return dp[N];
    }
}